### PR TITLE
Simplify GPS trace visibility levels

### PR DESCRIPTION
**Pull request by Rub21**
Ref:  https://github.com/openstreetmap/openstreetmap-website/issues/7089

This PR covers the GPS visibility simplification, new trace uploads now only accept `trackable` and `identifiable`. I removed the `private` and `public` options from the upload and edit forms. User can still change their own existing traces to `trackable` or `identifiable` whenever they want.

Deloyed here: https://gpsvisibility.apis.dev.openstreetmap.org/

### CODE REVIEW

**Comment by Rub21** [comment]
The visibility column keeps public as its [default](https://github.com/openstreetmap/openstreetmap-website/blob/master/db/structure.sql#L874):

```sql
visibility public.gpx_visibility_enum DEFAULT 'public'::public.gpx_visibility_enum NOT NULL
```

There is no migration yet. The app always sets the visibility in the controllers on upload, so this default never runs for new traces. We can change it to trackable in a separate PR, to keep this one small and easy to review

File: app/controllers/traces_controller.rb [hunk 1]
```diff
@@ -103,6 +103,9 @@ def edit
   def create
     @title = t ".upload_trace"
 
+    # New traces can only be trackable or identifiable.
+    return head :bad_request unless %w[trackable identifiable].include?(params[:trace][:visibility])
```
Comment (pablobm) [inline app/controllers/traces_controller.rb]: How about moving this question to the model?

```suggestion
    return head :bad_request unless Trace.valid_visibility?(params[:trace][:visibility])
```

Same in other instances.

File: app/controllers/traces_controller.rb [hunk 2]
```diff
@@ -211,15 +214,10 @@ def offline_redirect
...
+    return visibility if %w[trackable identifiable].include?(visibility)
 
-    if visibility
-      visibility.v
-    elsif current_user.preferences.find_by(:k => "gps.trace.public", :v => "default").nil?
-      "private"
-    else
-      "public"
-    end
+    "trackable"
```
Comment (pablobm) [inline app/controllers/traces_controller.rb]: Similarly, this perhaps should be `Trace.default_visibility`.

And since I'm at it: the whole concept of default visibility for a specific user could be moved to something like `User#default_trace_visibility`, similar to how there's `User#preferred_color_scheme` or `User#default_diary_language`.

File: app/views/traces/edit.html.erb [hunk 3]
```diff
@@ -9,9 +9,7 @@
   <%= f.text_field :description, :maxlength => 255 %>
   <%= f.text_field :tagstring %>
   <%= f.select :visibility,
-               [[t("traces.visibility.private"), "private"],
-                [t("traces.visibility.public"), "public"],
-                [t("traces.visibility.trackable"), "trackable"],
+               [[t("traces.visibility.trackable"), "trackable"],
                 [t("traces.visibility.identifiable"), "identifiable"]],
```
Comment (pablobm) [inline app/views/traces/edit.html.erb]: I think this list belongs in an appropriate model too 🙂 Or at least a helper. Then it can be used in both edit and new.

File: app/views/traces/edit.html.erb [hunk 4]
```diff
@@ -9,9 +9,7 @@
   <%= f.text_field :description, :maxlength => 255 %>
   <%= f.text_field :tagstring %>
   <%= f.select :visibility,
```
Comment (pablobm) [inline app/views/traces/edit.html.erb]: This is related to the failing test, so I'm guessing you are thinking about this, but just in case I'll phrase it here.

You say "User can still change their own existing traces to trackable or identifiable whenever they want", but this form risks doing it without the user noticing:

1. User has a pre-existing "private" trace.
2. User visits to edit the trace.
3. User changes the description, updates.
4. The trace is now "trackable" and the user may not have noticed.

Options that spring to mind:
1. This is fine.
2. If the trace has a deprecated visibility, add it to the options, then:
    a. the user can update the trace and keep the unsupported visibility unchanged.
    b. on submit we show an error. The user must change the visibility to a supported one if they want to edit the trace at all.

**Review by pablobm — COMMENTED** [review]
Is this PR ready or not? It's marked as WIP but not as draft, and a test is failing.

File: app/controllers/traces_controller.rb (hunk 1 above)
Comment (Rub21) [reply app/controllers/traces_controller.rb]: Good idea and best, I moved it to Trace.valid_visibility? and used it in the web and the API controller.

File: app/views/traces/edit.html.erb (hunk 3 above)
Comment (Rub21) [reply app/views/traces/edit.html.erb]: Done, I added a trace_visibility_options helper and both the new and edit forms use it now.

File: app/views/traces/edit.html.erb (hunk 4 above)
Comment (Rub21) [reply app/views/traces/edit.html.erb]: I went with option 2a. If the trace has an old visibility (private or public), the edit form now keeps it as the selected option, so updating the description doesn't change it. The visibility only changes if the user picks another value, I added a test for this too.

File: app/controllers/traces_controller.rb (hunk 2 above)
Comment (Rub21) [reply app/controllers/traces_controller.rb]: Good idea. I moved it to User#default_trace_visibility, following the default_diary_language pattern.

**Comment by Rub21** [comment]
Ready for review now, dropped the [WIP]. I addressed all the comments and fixed the failing test.

**Comment by Rub21** [comment]
I bumped the class-length limit to [350](https://github.com/Rub21/openstreetmap-website/blob/simplify-gps-visibility/.rubocop_todo.yml#L59) because adding default_trace_visibility pushed the User model over it.

File: app/models/trace.rb [hunk 5]
```diff
@@ -44,14 +44,28 @@ class Trace < ApplicationRecord
...
+  # Visibility values for new uploads, and the old ones kept for existing traces.
+  VISIBILITIES = %w[trackable identifiable].freeze
+  LEGACY_VISIBILITIES = %w[private public].freeze
+
   validates :user, :associated => true
   validates :name, :presence => true, :length => 1..255, :characters => true
   validates :description, :presence => { :on => :create }, :length => 1..255, :characters => true
   validates :timestamp, :presence => true
-  validates :visibility, :inclusion => %w[private public trackable identifiable]
+  validates :visibility, :inclusion => VISIBILITIES + LEGACY_VISIBILITIES
```
Comment (pablobm) [inline app/models/trace.rb:55]: Thoughts on a custom validator here? To check that there's no accidental jump from valid to legacy value.

**Review by pablobm — COMMENTED** [review]
This also needs a migration: the DB schema still has "public" as default value for the visibility column.

File: app/models/trace.rb (hunk 5 above)
Comment (Rub21) [reply app/models/trace.rb:55]: I added one. There's now a visibility_not_changed_to_legacy validator that runs on update. It only blocks a trace that moves from a valid value (trackable/identifiable) back to a legacy one (public/private):

**Comment by Rub21** [comment]
This PR is ready to review.

For new uploads you can test on the official dev deployment: https://gpsvisibility.apis.dev.openstreetmap.org/. There you can check that uploads only offer trackable and identifiable.

What that stack can't cover is the existing-data case, because it builds with a clean database. So I couldn't reproduce the real situation where a user already has legacy traces that need updating. For that I seeded my own DB with traces in all four visibilities (public, private, trackable, identifiable) and ran the migration on top of it:

https://simplify-gps-visibility.167.233.138.142.nip.io/

On that instance you can check that when you edit an old trace you can move it to a current value but not back to a legacy one (public/private). It's a throwaway instance for testing — ping me if you need login credentials.

File: app/models/user.rb [hunk 6]
```diff
@@ -294,6 +294,15 @@ def default_diary_language=(language)
     preference.update!(:v => language)
   end
 
+  def default_trace_visibility
+    visibility = preferences.find_by(:k => "gps.trace.visibility")&.v
+    if Trace.valid_visibility?(visibility)
+      visibility
+    else
+      "trackable"
```
Comment (pablobm) [inline app/models/user.rb]: I'd put this in `Trace.default_visibility` or similar.

File: test/controllers/api/traces_controller_test.rb [hunk 7]
```diff
@@ -156,39 +156,20 @@ def test_create
...
-      assert_equal %w[new trace], trace.tags.order(:tag).collect(&:tag)
-      assert_equal "private", trace.visibility
-      assert_not trace.inserted
-      assert_equal File.new(fixture).read, trace.file.blob.download
-      trace.destroy
-      assert_equal "private", second_user.preferences.find_by(:k => "gps.trace.visibility").v
+      assert_no_difference "Trace.count" do
+        post api_traces_path, :params => { :file => file, :description => "New Trace", :tags => "new,trace", :public => 0 }, :headers => auth_header
+      end
+      assert_response :bad_request
```
Comment (pablobm) [inline test/controllers/api/traces_controller_test.rb:185]: Since we are changing this, how about moving it to a separate test? `test_create_legacy_public_flag` or something like that. I find these long tests difficult to follow (although it already helped that you shortened it significantly).

Also how about a test for a legacy visibility? Just one is enough, no need for both. Also in a separate test if possible: `test_create_legacy_visibility`.

File: test/controllers/traces_controller_test.rb [hunk 8]
```diff
@@ -427,6 +430,15 @@ def test_create_post
...
     assert_equal "trackable", user.preferences.find_by(:k => "gps.trace.visibility").v
+
+    # Rewind the file
+    file.rewind
+
+    # The legacy public and private values are no longer accepted on upload
+    assert_no_difference "Trace.count" do
+      post traces_path, :params => { :trace => { :gpx_file => file, :description => "New Trace", :tagstring => "new,trace", :visibility => "public" } }
+    end
+    assert_response :bad_request
```
Comment (pablobm) [inline test/controllers/traces_controller_test.rb]: Again, a good opportunity to put this in a separate test.

File: app/controllers/api/traces_controller.rb [hunk 9]
```diff
@@ -28,13 +28,8 @@ def create
...
 
-      if visibility.nil?
-        visibility = if params.fetch(:public, "0").to_i.nonzero?
-                       "public"
-                     else
-                       "private"
-                     end
-      end
+      # New traces can only be trackable or identifiable.
+      return head :bad_request unless Trace.valid_visibility?(visibility)
```
Comment (pablobm) [inline app/controllers/api/traces_controller.rb]: Actually, this is probably best moved to an `:on => :create` validation. Also avoids repeating it on the non-API controller.

File: app/models/user.rb (hunk 6 above)
Comment (Rub21) [reply app/models/user.rb]: Done, i have added Trace.default_visibility and default_trace_visibility now uses it as the fallback,  the value lives in the Trace model.

File: app/controllers/api/traces_controller.rb (hunk 9 above)
Comment (Rub21) [reply app/controllers/api/traces_controller.rb]: Done. I moved it to a :on => :create validation in the model and removed the check from both controllers. The API still returns 400, and the web form now re-renders with the error if a legacy visibility is passed. That won't normally happen though, since the form only offers the valid visibilities.

File: test/controllers/api/traces_controller_test.rb (hunk 7 above)
Comment (Rub21) [reply test/controllers/api/traces_controller_test.rb:185]: I have split  it into test_create_legacy_public_flag for the old public flag, and added test_create_legacy_visibility for a legacy visibility value. Just one (public), as you said.

File: test/controllers/traces_controller_test.rb (hunk 8 above)
Comment (Rub21) [reply test/controllers/traces_controller_test.rb]: Done., I have moved it to its own test, test_create_post_with_legacy_visibility.

File: app/controllers/api/traces_controller.rb (hunk 9 above)
Comment (Rub21) [reply app/controllers/api/traces_controller.rb]: Quick update. The controller check was the original version here, and all tests passed with it.

```
      # New traces can only be trackable or identifiable.
      return head :bad_request unless Trace.valid_visibility?(visibility)
```

I tried moving it to a model validation like you suggested, but it breaks the legacy tests. The factory creates public/private traces to set up existing-data cases, and the validation blocks that on create. Rewriting those tests would not be a good option. Some of them test legacy behavior on purpose, and public and private are still valid values in the database for existing traces.


So I would keep the check in the controllers, calling Trace.valid_visibility?. The rule is really about the upload, and a trace can still hold a legacy value that exists in the database. What do you think?

Comment (pablobm) [reply app/controllers/api/traces_controller.rb]: My concern is that then we leave the tests littered with references to legacy values, giving the impression that they are the normal case.

I think the tests need to be converted to use the new values, while still acknowledging that legacy values will remain in DB and need to be handled appropriately.

These are the steps to get there:
- Generally, convert instances of public/private to identifiable/trackable...
- ..except where it makes sense to test legacy values. This might be existing tests (eg: adding records to `test_index`) or new ones that exist specifically to test these edge cases.
- When the tests need legacy values, they can be created with this:
```
# Source - https://stackoverflow.com/a/25391890
# Posted by Jason Denney, modified by community. See post 'Timeline' for change history
# Retrieved 2026-07-02, License - CC BY-SA 4.0

trait :without_validations do
  to_create { |instance| instance.save(validate: false) }
end
```

If you put these changes in the commit titled "Validate the create visibility in the model", I want to think it'll make the PR easier to review. Hopefully it will tuck all those changes away in a commit with a clear intent.

Comment (Rub21) [reply app/controllers/api/traces_controller.rb]: Done!

 I moved the tests to the new values (trackable/identifiable). I kept public/private only where the test is really about old traces: editing one, keeping its value, and the visible_to scopes.

For those I added a without_validations trait to the factory, so a test can still insert a legacy trace even though new uploads reject it:

```rb
trait :without_validations do
  to_create { |instance| instance.save(:validate => false) }
end
```
I also updated test_validations to expect public/private rejected on create. It's all in the "Validate the create visibility in the model" commit.

File: test/controllers/api/tracepoints_controller_test.rb [hunk 10]
```diff
@@ -28,7 +28,7 @@ def test_routes
     end
 
     def test_tracepoints
```
Comment (pablobm) [inline test/controllers/api/tracepoints_controller_test.rb]: I think worth renaming as it's a legacy value, and it can confuse readers as to what exactly is being tested here (I was confused), and why a legacy value appears to be a "default" in the sense that the test name doesn't say anything special.

```suggestion
    def test_tracepoints_public
```

File: app/helpers/trace_helper.rb [hunk 11]
```diff
@@ -18,4 +18,15 @@ def trace_icon(trace, options = {})
...
+  # Options for the visibility dropdown. Keeps an old value in the list so
+  # editing the trace does not change it by mistake.
+  def trace_visibility_options(trace)
+    visibilities = Trace::VISIBILITIES.dup
+    # If the trace still has an old visibility (private or public), add it so it stays selected.
+    visibilities.unshift(trace.visibility) if Trace.legacy_visibility?(trace.visibility)
+    visibilities.map do |visibility|
+      [t("traces.visibility.#{visibility}"), visibility]
+    end
+  end
```
Comment (pablobm) [inline app/helpers/trace_helper.rb:26]: Instead of the comments, I would provide a test. And since `test/helpers/trace_helper_test.rb` would be a new test file, you can use the `test "long description"` form to make it more readable.

**Review by pablobm — COMMENTED** [review]
Almost there! 🤞

File: test/controllers/api/tracepoints_controller_test.rb (hunk 10 above)
Comment (Rub21) [reply test/controllers/api/tracepoints_controller_test.rb]: Thanks, I renamed the function to test_tracepoints_public

File: app/helpers/trace_helper.rb (hunk 11 above)
Comment (Rub21) [reply app/helpers/trace_helper.rb:26]: Done, I've removed the comments and added test/helpers/trace_helper_test.rb using the test "..." form, covering both the normal and the legacy trace cases.

**Comment by Rub21** [comment]
I fixed the conflict in `db/structure.sql` and squashed the commits to keep the history readable.

File: app/models/trace.rb [hunk 12]
```diff
@@ -14,7 +14,7 @@
 #  timestamp   :datetime         not null
 #  description :string           default(""), not null
 #  inserted    :boolean          not null
-#  visibility  :enum             default("public"), not null
+#  visibility  :enum             default("trackable"), not null
```
Comment (tomhughes) [inline app/models/trace.rb:17]: There's something mixed up in your commit history - this comment is changing in the first commit before the database migration that would trigger it has been added.

File: app/helpers/trace_helper.rb [hunk 13]
```diff
@@ -18,4 +18,12 @@ def trace_icon(trace, options = {})
...
     trace_image(trace, :animated => true, :size => 250, **options)
   end
+
+  def trace_visibility_options(trace)
+    visibilities = Trace::VISIBILITIES.dup
+    visibilities.unshift(trace.visibility) if Trace.legacy_visibility?(trace.visibility)
+    visibilities.map do |visibility|
+      [t("traces.visibility.#{visibility}"), visibility]
+    end
+  end
```
Comment (tomhughes) [inline app/helpers/trace_helper.rb:26]: Might this be better as a method on the Trace model? Given it is mostly accessing attributes of the model and a list of visibilities from the model and it's value is deliberately varying with the details of the model record?

Also this function has some comments added in the first commit that you then remove in the third commit which seems pointless.

File: test/controllers/api/traces_controller_test.rb [hunk 14]
```diff
@@ -163,50 +163,46 @@ def test_create
...
+      fixture = Rails.root.join("test/gpx/fixtures/a.gpx")
+      file = Rack::Test::UploadedFile.new(fixture, "application/gpx+xml")
+      user = create(:user)
+      auth_header = bearer_authorization_header user
+
+      assert_no_difference "Trace.count" do
+        post api_traces_path, :params => { :file => file, :description => "New Trace", :tags => "new,trace", :visibility => "public" }, :headers => auth_header
+      end
+      assert_response :bad_request
     end
```
Comment (tomhughes) [inline test/controllers/api/traces_controller_test.rb:206]: Shouldn't this test both legacy visibilities not just one of them? Note that this varies over time through the commits but in the first commit it is initially added as "public and private no longer supported" but only tests public.

File: app/models/trace.rb [hunk 15]
```diff
@@ -44,14 +44,35 @@ class Trace < ApplicationRecord
...
+  VISIBILITIES = %w[trackable identifiable].freeze
+  LEGACY_VISIBILITIES = %w[private public].freeze
+
   validates :user, :associated => true
   validates :name, :presence => true, :length => 1..255, :characters => true
   validates :description, :presence => { :on => :create }, :length => 1..255, :characters => true
   validates :timestamp, :presence => true
-  validates :visibility, :inclusion => %w[private public trackable identifiable]
+  validates :visibility, :inclusion => VISIBILITIES + LEGACY_VISIBILITIES
+  validates :visibility, :inclusion => { :in => VISIBILITIES }, :on => :create
```
Comment (tomhughes) [inline app/models/trace.rb:56]: This is added in the second commit but wouldn't it be better placed in the first?

You also then wouldn't need to add a check in the controller in the first commit only to remove it again in the second.

**Review by tomhughes — COMMENTED** [review]
I suspect the second commit, to change the default visibility, would be better done first before you start trying to deprecate the public and private visibilities.

File: app/models/trace.rb (hunk 12 above)
Comment (Rub21) [reply app/models/trace.rb:17]: Fixed, I reordered the commits so the schema comment changes in the same commit as the migration.

File: app/helpers/trace_helper.rb (hunk 13 above)
Comment (Rub21) [reply app/helpers/trace_helper.rb:26]: Done, I moved the list logic to a new `selectable_visibilities` method  in the  Trace model, and the helper only maps the values to translated labels for the form.

File: app/models/trace.rb (hunk 15 above)
Comment (Rub21) [reply app/models/trace.rb:56]: Done, the create validation is now in the `Limit new trace uploads` commit from the start, so the temporary controller check is gone from the history.

File: test/controllers/api/traces_controller_test.rb (hunk 14 above)
Comment (Rub21) [reply test/controllers/api/traces_controller_test.rb:206]: Done, the test now checks both public and private. The model and helper tests also cover both legacy values now.

**Comment by Rub21** [comment]
Done,I reordered the commits and  the default visibility change is the first commit now. I've also folded the test-only commits into the main one, so the PR is two commits now.

**Review by tomhughes — APPROVED** [review]
I think this looks good now - the only questions now are around if/when we deploy this and the public messaging that will be needed around the change.

File: test/helpers/trace_helper_test.rb [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  test "visibility options pair translated labels with the selectable visibilities" do
+    %w[public private].each do |legacy_visibility|
+      trace = build(:trace, :visibility => legacy_visibility)
+      options = trace_visibility_options(trace)
+      assert_equal trace.selectable_visibilities, options.map(&:last)
+      options.each do |label, visibility|
+        assert_equal t("traces.visibility.#{visibility}"), label
+      end
+    end
+  end
```
Comment (pablobm) [inline test/helpers/trace_helper_test.rb]: I tend to avoid this style of testing:
- I have been bitten in the past by clever loops that allowed me to test every single combination, but then at some point the list was empty, the loop didn't actually run, and nothing was tested.
- The loops also make it a bit difficult to follow.
- The test is repeating code from the tested method. For example, it compares `trace.selectable_visibilities` to the output, but that's exactly where it's generated from, so obviously it'll be the same.

I would simplify here, repeating code if needs be, and not worrying about the translated string as it's less important anyway:

```suggestion
  test "trace_visibility_options includes a legacy option for traces that have it" do
    public_trace = build(:trace, :visibility => "public")
    options = trace_visibility_options(public_trace)
    assert_equal %w[public trackable identifiable], options.map(&:last)

    private_trace = build(:trace, :visibility => "private")
    options = trace_visibility_options(private_trace)
    assert_equal %w[private trackable identifiable], options.map(&:last)
  end
```

Since I was at it, I also rephrased the description.

Comment (Rub21) [reply test/helpers/trace_helper_test.rb]: Thanks, i've applied your suggestion and I also added a small test for the non-legacy case.

**Comment by Rub21** [comment]
>  the only questions now are around if/when we deploy this and the public messaging that will be needed around the change.

Minh and I already discussed this and prepared a [doc with a draft announcement ](https://hackmd.io/@Rub21/H1MFlSLVMl) plus a list of places  where to update and notify. For the deploy timing, we can discuss it in the next Ops meeting so everyone is in sync.

File: test/helpers/trace_helper_test.rb (hunk 16 above)
Comment (pablobm) [reply test/helpers/trace_helper_test.rb]: Nice, thank you :muscle:

**Review by pablobm — APPROVED** [review]

**Comment by 1ec5** [comment]
[Forum announcement](https://community.openstreetmap.org/t/changes-coming-to-gps-trace-uploads/145728)

**Comment by HolgerJeromin** [comment]
~~A clean approach would be to add this to the capabillity api, so programs can be adjusted now.
But depending on the timescale this could be late anyway.~~

But perhaps adjusting every program now is the right thing to do, because the data will be migrated anyway.

**Comment by pablobm** [comment]
@HolgerJeromin, what do you mean exactly here?

> the data will be migrated anyway

If you mean that entries in the DB will be transformed from public/private to identifiable/trackable, then that is not going to happen.

**Comment by tomhughes** [comment]
We certainly won't be changing the status of any existing tracks, though obviously the owners of those tracks are free to do so.

At some point those legacy tracks will probably be excluded from the API point cloud download in order to make it more efficient.

**Comment by HolgerJeromin** [comment]
> the data will be migrated anyway

Better term would be: 

Reducing the options in the apps can now be happen without a problem for the future.

**Comment by tomhughes** [comment]
We discussed this on the operations call last week and we think we're ready to deploy it now. Thanks for all the work.
